File: tests/fixtures/ldap_injection/typescript/unsafe_ldap_search.ts
// Unsafe: ldapjs `client.search` receives a filter assembled from req.query.
// Bound-variable idiom: the closure-captured `client` carries
// `TypeKind::LdapClient` (forwarded from the top-level body to the function
// body by `taint::inject_external_type_facts`), so type-qualified receiver
// resolution rewrites `client.search` → `LdapClient.search`.
import * as ldap from 'ldapjs';

const client = ldap.createClient({ url: 'ldap://example.com' });

export function lookup(req: any, res: any) {
    const user = req.query.user;
    const filter = '(uid=' + user + ')';
    client.search('ou=people,dc=example,dc=com', { filter: filter }, (err: any) => { res.json({ ok: !err }); });
}

File: tests/fixtures/ldap_injection/python/unsafe_ldap_search.py
# Unsafe: tainted form data concatenated into an LDAP filter and passed to
# python-ldap's `search_s`.  The bound receiver `conn` is typed as LdapClient
# via `ldap.initialize`, and the suffix matcher on `search_s` also catches the
# call directly.
import ldap
from flask import request


def lookup():
    conn = ldap.initialize("ldap://example.com")
    user = request.form["user"]
    flt = "(uid=" + user + ")"
    return conn.search_s("ou=people,dc=example,dc=com", ldap.SCOPE_SUBTREE, flt)
